### PR TITLE
perf(text): avoid avoidable O(n) copies/scans in injection/formatting/offset paths

### PR DESCRIPTION
**Pull request by atusy**
## Summary
- `injection.rs` / `formatting.rs`: replace `text().to_string()` with `text_arc()` (refcount bump) at two read-only call sites; the Concatenated-formatting pipeline now only pays for an owned `String` when it actually mutates text.
- `offset_calculator.rs`: `apply_offset_to_position` (`#offset!` row handling) now walks only the span between the byte position and the target line, bounded by the row-offset distance, instead of two full-document scans. Pinned the existing overshoot-clamps-to-last-line behavior with a new test.
- Skipped the `text_sync.rs` per-change `PositionMapper` rebuild sub-item: the suggested descending-sort/splice fix changes behavior under LSP's sequential content-change semantics (a later range in a batch can depend on line shifts from an earlier one). Commented on the issue with a counterexample; left as a follow-up candidate.

Closes part of #581 (item 2 remains open, see issue comment).

## Test plan
- [x] `cargo test --lib` (2461 passed)
- [x] `cargo clippy --all-targets --all-features -- -D warnings`
- [x] `cargo fmt --check`
- [x] full e2e suite via pre-commit gate

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved position/line offset handling: overshoots clamp to the last line start, and offset computations now safely handle byte positions that land mid–multi-byte character.
* **Performance**
  * Reduced unnecessary text copying during formatting and injection resolution by reusing shared document text snapshots.
  * Optimized offset calculations to work relative to nearby line movement rather than scanning the full document.
* **Tests**
  * Added edge-case coverage for clamping and non-panicking behavior with mid-character starting positions.

### CODE REVIEW

File: src/analysis/offset_calculator.rs [hunk 1]
```diff
@@ -86,59 +92,37 @@ fn apply_offset_to_position(
...
-            }
-            if ch == '\n' {
-                lines_seen += 1;
-                target_line_start = line_start_byte + i + 1;
+    // Start of the line containing byte_pos, found by scanning backward from
+    // byte_pos rather than forward from the document start.
+    let current_line_start = text[..byte_pos.min(text.len())]
+        .rfind('\n')
+        .map(|i| i + 1)
+        .unwrap_or(0);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/offset_calculator.rs]: Fixed in 29c49ef8d — snapped byte_pos to a char boundary with floor_char_boundary before the initial slice. Good catch, thanks.

File: src/analysis/offset_calculator.rs [hunk 2]
```diff
@@ -86,59 +92,37 @@ fn apply_offset_to_position(
...
+        let mut pos = current_line_start;
+        for _ in 0..row_offset.unsigned_abs() {
+            if pos == 0 {
                 break;
             }
-            if ch == '\n' {
-                lines_seen += 1;
-            }
+            pos = text[..pos - 1].rfind('\n').map(|i| i + 1).unwrap_or(0);
         }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/offset_calculator.rs]: Fixed in 29c49ef8d — same fix as the Copilot comment on this line: snapped byte_pos to a char boundary with floor_char_boundary before the initial slice, and added a regression test with a mid-codepoint byte_pos.

File: src/analysis/offset_calculator.rs [hunk 3]
```diff
@@ -86,59 +92,41 @@ fn apply_offset_to_position(
...
-            if ch == '\n' {
-                lines_seen += 1;
-                target_line_start = line_start_byte + i + 1;
+    // Start of the line containing byte_pos, found by scanning backward from
+    // byte_pos rather than forward from the document start. Snap to a char
+    // boundary first: unlike the char_indices()-based scan this replaced,
+    // slicing at a raw byte_pos panics if it lands mid-codepoint (e.g. a
+    // stale tree-sitter offset) — the caller's later clamp/snap only applies
+    // to the *returned* value, not this internal slice.
+    let current_line_start = text[..floor_char_boundary(text, byte_pos.min(text.len()))]
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/offset_calculator.rs]: Fixed — dropped the redundant .min(text.len()), floor_char_boundary already clamps.

File: src/lsp/lsp_impl/text_document/formatting.rs [hunk 4]
```diff
@@ -168,13 +168,13 @@ impl Kakehashi {
...
                 // warning rather than formatting against stale regions.
-                let mut current = original.clone();
+                let mut current = original.to_string();
                 let mut producers = 0usize;
                 let mut sole_edits: Option<Vec<TextEdit>> = None;
                 for layer in &layer_cfg.priorities {
                     let edits = match layer {
                         LayerSource::Virt => {
-                            if current != original {
+                            if current.as_str() != &*original {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/text_document/formatting.rs]: Good suggestion — implemented with Cow<str> as you outlined (Borrowed from original, Owned once apply_text_edits actually produces a mutation).
